request.go

```diff
@@ -212,66 +222,111 @@ func (exo *Client) parseResponse(resp *http.Response) (json.RawMessage, error) {
-	body, err := exo.request(req.name(), req)
+func (exo *Client) processAsyncJob(ctx context.Context, job *asyncJob) {
+	defer close(job.responseChan)
+	defer close(job.errorChan)
+
+	body, err := exo.request(job.command.name(), job.command)
 	if err != nil {
-		return nil, err
+		job.errorChan <- err
+		return
 	}
 
-	// Is it a Job?
-	job := new(JobResultResponse)
-	if err := json.Unmarshal(body, &job); err != nil {
-		return nil, err
+	jobResult := new(AsyncJobResult)
+	if err := json.Unmarshal(body, jobResult); err != nil {
+		r := new(ErrorResponse)
+		if e := json.Unmarshal(body, r); e != nil {
+			job.errorChan <- r
+			return
+		}
+		job.errorChan <- err
+		return
 	}
 
-	// Error response
-	errorResponse := new(ErrorResponse)
 	// Successful response
-	resp := req.asyncResponse()
-	if job.JobID == "" || job.JobStatus != Pending {
-		if err := json.Unmarshal(*job.JobResult, resp); err != nil {
-			return job, err
-		}
-		return resp, nil
+	if jobResult.JobID == "" || jobResult.JobStatus != Pending {
+		job.responseChan <- jobResult
+		return
 	}
 
-	// we've got a pending job
-	result := &QueryAsyncJobResultResponse{
-		JobStatus: job.JobStatus,
-	}
-	for async.Retries > 0 && result.JobStatus == Pending {
-		time.Sleep(time.Duration(async.Delay) * time.Second)
+	for job.retries > 0 {
+		select {
+		case <-ctx.Done():
+			job.errorChan <- ctx.Err()
+			return
+		default:
+			job.retries--
 
-		async.Retries--
+			end, _ := ctx.Deadline()
 
-		req := &QueryAsyncJobResult{JobID: job.JobID}
-		resp, err := exo.Request(req)
-		if err != nil {
-			return nil, err
-		}
-		result = resp.(*QueryAsyncJobResultResponse)
-	}
+			when := end.Add(time.Duration(job.delay*-job.retries) * time.Second)
+			time.Sleep(when.Sub(time.Now()))
 
-	if result.JobStatus == Failure {
-		if err := json.Unmarshal(*result.JobResult, &errorResponse); err != nil {
-			return nil, err
+			req := &QueryAsyncJobResult{JobID: jobResult.JobID}
+			resp, err := exo.Request(req)
+			if err != nil {
+				job.errorChan <- err
+				return
+			}
+
+			result := resp.(*QueryAsyncJobResultResponse)
+			if result.JobStatus == Success {
+				job.responseChan <- (*AsyncJobResult)(result)
+				return
+			} else if result.JobStatus == Failure {
+				r := new(ErrorResponse)
+				e := json.Unmarshal(*result.JobResult, r)
+				if e != nil {
+					job.errorChan <- e
+					return
+				}
+				job.errorChan <- r
+				return
+			}
 		}
-		return errorResponse, errorResponse
 	}
 
-	if result.JobStatus == Pending {
-		return result, fmt.Errorf("Maximum number of retries reached")
-	}
+	job.errorChan <- fmt.Errorf("Maximum number of retries reached")
+}
 
-	if err := json.Unmarshal(*result.JobResult, resp); err != nil {
-		if err := json.Unmarshal(*result.JobResult, errorResponse); err != nil {
+// AsyncRequest performs an asynchronous request and polls it for retries * day [s]
+func (exo *Client) AsyncRequest(req AsyncCommand, async AsyncInfo) (interface{}, error) {
```

Comment:
Same here, `AsyncRequestWithContext(ctx, ...)`

Reply:
I'm adding this on the todo list. #42